Reorg TLV handling and add test sequence for SMP TLVs

eighthave opened the pull request:
This pull request is a Work In Progress, but I wanted to post it because I'm a bit stumped why the SMP test I wrote does not complete.  So I'm hoping someone can spot the problem.  Then I can fix it, rebase, and then this request will be ready for merging.

I wanted to write the SMP tests so we can reorganize the TLV handling and have something to make sure we didn't break things.  I already included one commit that reorgs the TLV handling and fixes #11 also.

I also purged some dead code, and added some javadoc comments on the public `interface`.

eighthave committed:
set up working logging for tests

It outputs a lot now, but it is easy to change the levels.  This took waaay
too long to figure out.

eighthave committed:
purge obsolete OtrEngineHost methods from DummyOtrEngineHostImpl

just cleaning house here...  It seems that OtrEngineHost at one point provided
a lot more methods.  This commit purges a bunch of methods from
DummyOtrEngineHostImpl that were neither used in the code, nor listed in the
OtrEngineHost interface.

eighthave committed:
improve OtrKeyManagerImplTest to also test unverify and re-reading file

eighthave committed:
add some javadoc to OtrKeyManager and OtrEngineHost interfaces

eighthave committed:
break out SMP TLVs into separate functions to make adding more TLVs clean

It was not very easy to follow how the TLVs where broken out before, so
this commit breaks out all of the TLVs in a single case statement, then
handles each TLV in its own method.  This will make it cleaner to add more
TLVs like TLV8 and OTRDATA.

Also, this can change the resulting output from a Data Message because this
does not force the message contents to null if the Data Message contains a
SMP TLV.  The OTR spec is pretty clear about this, saying: "...a human-
readable message..., optionally followed by... zero or more TLV
(type/length/value) records..."

libotr and pure-python-otr both will output any message content in a Data
Message that contains any TLV, for example, here's pure-python-potr:
https://github.com/python-otr/pure-python-otr/blob/1.0.1/src/potr/crypt.py#L190

closes https://github.com/otr4j/otr4j/issues/11

eighthave committed:
test the SMP (Socialist Millionaire Protocol) sequence

Make sure that the SMP sequence is working, both verifying correctly with
passwords, and failing properly with bad passwords.

cobratbq commented:
hi @eighthave I'll have a look at this this weekend. Then I'll have some more time to actually dive into the changes.

cobratbq commented:
@eighthave, not cool man! You are generating a new keypair on every call to getLocalKeyPair()! (See DummyOtrEngineHostImpl)

Try something like this, that should fix your problem :+1: 

``` java
        private HashMap<SessionID, KeyPair> keypairs = new HashMap<SessionID, KeyPair>();

        public KeyPair getLocalKeyPair(SessionID paramSessionID) {
            KeyPair keypair = this.keypairs.get(paramSessionID);
            if (keypair == null) {
                KeyPairGenerator kg;
                try {
                    kg = KeyPairGenerator.getInstance("DSA");
                    keypair = kg.genKeyPair();
                    this.keypairs.put(paramSessionID, keypair);
                } catch (NoSuchAlgorithmException ex) {
                    logger.log(Level.SEVERE, null, ex);
                }
            }
            return keypair;
        }
```

On the positive side: your test succeeded immediately after having fixed that mistake :-D

EDIT: Okay, technically it wasn't your code that keeps generating new keypairs. Still, it was a really annoying problem :-P

cobratbq committed:
DummyClient: store local key pairs to support more elaborate tests

Previously, DummyOtrEngineHostImpl would generate a new local keypair on
every call to getLocalKeyPair(), which made more real-life tests impossible.
This changes to stored each generated local key pair for the life of the
DummyClient instance to support tests that can have full OTR sessions with
conversations.

https://github.com/otr4j/otr4j/pull/16#issuecomment-74430153

eighthave committed:
SMPTest: include unicode tests for both question/answer and password

Since SMP verification is based on user input, it should handle all manner
of unicode that anything might throw at it.  It should also handle badly
formed strings.  The downside is that this test suite takes 5 minutes to
run because of the whole OTR and SMP setup for each test session.

Some of these are failing, it seems that the question handling does not
properly sanitize the input, or if it is meant to accept special chars,
then it is not handling them correctly.

refs https://github.com/otr4j/otr4j/issues/30
closes https://github.com/otr4j/otr4j/pull/16

eighthave commented:
Ok, this is now complete, thanks to @cobratbq for the final tip that got the whole thing working.  I also added some more SMP tests that throw lots of unicode at the process.  The unicode chars seem to work fine, but it seems that special chars like `\n \t \r \0` in the question cause an exception, so I filed a bug: #30.

cobratbq commented on pom.xml:
```diff
@@ -104,6 +104,9 @@
                             %4$-6s %2$s(): %3$s "%5$s%6$s"%n</java.util.logging.SimpleFormatter.format>
+                  <java.util.logging.config.file>
+                    ${basedir}/logging.properties
+                  </java.util.logging.config.file>
```
Fix indentation

eighthave replied:
fixed

cobratbq commented on src/main/java/net/java/otr4j/session/Session.java:
```diff
@@ -648,12 +648,29 @@ private String handleDataMessage(DataMessage data) throws OtrException {
                     for (TLV tlv : tlvs) {
                         switch (tlv.getType()) {
-                            case TLV.DISCONNECTED:
+                            // TVL.PADDING // TLV0
```
I think we should make this case explicit. So TLV.PADDING: break; That way, if ever we decide to have a "default" case for unsupported types, we can be sure that we _only_ catch the unsupported types.

eighthave replied:
makes sense, fixed.  I also added a `default:` to throw a warning message to aid debugging.

cobratbq commented:
I haven't checked the various processTlv\* methods yet. I may be able to do that tomorrow. Or maybe languitar beats me to it, which is fine too of course.

eighthave commented:
the various `processTlv`\* methods are just the unmodified code from before, but instead of being in an if-then-else ladder, they are individual methods.

cobratbq commented:
Indeed, processTlv\* are just split into separate methods. I checked the rest, including a quick test use of the otr4j library and everything's in order AFAICT. Merging ...
